src/Yoti/Util/Profile/AnchorConverter.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace Yoti\Util\Profile;
+
+use Traversable;
+use phpseclib\File\ASN1;
+use phpseclib\File\X509;
+use Attrpubapi_v1\Anchor;
+use Yoti\Entity\Anchor as YotiAnchor;
+
+class AnchorConverter
+{
+    public function __construct()
+    {
+        $this->ASN1 = new ASN1();
+        $this->X509 = new X509();
+    }
+
+    /**
+     * Convert Protobuf Anchors to a map of Yoti Anchors
+     * And merge the result into the reference array
+     *
+     * @param Anchor $anchor
+     * @param $yotiAnchorsMap
+     */
+    public function convertToYotiAnchors(Anchor $anchor, &$yotiAnchorsMap)
+    {
+        $anchorSubType = $anchor->getSubType();
+        $certificateList = $anchor->getOriginServerCerts();
+        $yotiSignedTimeStamp = $this->convertToYotiSignedTimestamp($anchor);
+        $X509CertsList = $this->convertCertsListToX509($anchor->getOriginServerCerts());
+
+        foreach ($certificateList as $certificate) {
+            $certObj = $this->convertCertToX509($certificate);
+            $certExtArr = $certObj->tbsCertificate->extensions;
+
+            if (count($certExtArr) > 1) {
+                $anchorType = $certExtArr[1]->extnId;
+                $extEncodedValue = $certExtArr[1]->extnValue;
+
+                if ($decodedAnchorValue = $this->decodedAnchorValue($extEncodedValue)) {
+                    $yotiAnchorsMap[$anchorType][] = new YotiAnchor(
+                        $decodedAnchorValue,
+                        $anchorType,
+                        $anchorSubType,
+                        $yotiSignedTimeStamp,
+                        $X509CertsList
+                    );
+                }
+            }
+        }
+    }
+
+    /**
+     * @param string $extEncodedValue
+     *
+     * @return null
+     */
+    private function decodedAnchorValue($extEncodedValue)
+    {
+        $encodedBER = $this->X509->_extractBER($extEncodedValue);
+        $decodedValArr = $this->ASN1->decodeBER($encodedBER);
+        if (isset($decodedValArr[0]['content'][0]['content'])) {
+            return $decodedValArr[0]['content'][0]['content'];
+        }
+        return NULL;
+    }
+
+    /**
+     * @param \Attrpubapi_v1\Anchor $anchor
+     *
+     * @return \Yoti\Entity\SignedTimeStamp
+     */
+    private function convertToYotiSignedTimestamp(Anchor $anchor)
+    {
+        $signedTimeStamp = new \Compubapi_v1\SignedTimestamp();
+        $signedTimeStamp->mergeFromString($anchor->getSignedTimeStamp());
+
+        $timeInSeconds = round($signedTimeStamp->getTimestamp()/1000000);
```

Comment:
The timestamp we return has to be accurate to the microsecond.
Also, what TimeZone will be set on this?  Everything has to be UTC.

Java _time_ classes don't do microsecond accuracy so we had to create a new object for it.  That was a pain, but does have the advantage that we don't have to consider time zones anymore - we simply never included a timezone on it and documented that everything is UTC.  We don't use any of _time_ classes, so we don't have to worry about time zone conversion.

Reply:
Timestamp seems to be in UTC `12-07-2017 20:24:40 UTC`
Will double check anyway.